## Change the file directory if you run this file on your own laptop.

In [1]:
import pandas as pd

In [6]:
patients = pd.read_csv("data/patients.csv")
print(patients.columns)
patients

Index(['Id', 'BIRTHDATE', 'DEATHDATE', 'SSN', 'DRIVERS', 'PASSPORT', 'PREFIX',
       'FIRST', 'LAST', 'SUFFIX', 'MAIDEN', 'MARITAL', 'RACE', 'ETHNICITY',
       'GENDER', 'BIRTHPLACE', 'ADDRESS', 'CITY', 'STATE', 'COUNTY', 'FIPS',
       'ZIP', 'LAT', 'LON', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE',
       'INCOME'],
      dtype='object')


Id   BIRTHDATE   DEATHDATE  \
0     bb30994a-ad91-d617-2466-e9ef4c3838dd  2010-10-24         NaN   
1     22f309fb-7b30-144b-cedb-ce932fe286e4  2009-03-16         NaN   
2     343787c0-ee04-90d8-61a5-4ed996438c63  1981-07-08         NaN   
3     162a2c8c-f58e-95e5-d952-13b08627fd1a  1978-03-07         NaN   
4     a1c82b81-cc4e-1eef-1046-a4081a4b12b4  1991-02-12         NaN   
...                                    ...         ...         ...   
5710  f4cee386-514c-de23-bf20-6d88f3153571  1986-07-19         NaN   
5711  7a9d48f0-623c-9dda-1de7-681f1d2e7f7b  1958-07-20  2015-04-09   
5712  27d948ff-e081-99c7-83a3-ef9e94a6b01d  2000-10-17         NaN   
5713  bd7ebd2a-6e12-2b09-a22f-a3ed83a93561  1958-07-20         NaN   
5714  30bfbb3e-2f6a-f599-c87e-3b3fdf0d0b33  1946-10-19         NaN   

              SSN    DRIVERS    PASSPORT PREFIX        FIRST           LAST  \
0     999-25-7461        NaN         NaN    NaN    Nathan164    Schuster709   
1     999-37-4271        NaN         NaN    NaN      Asia282       Lesch175   
2     999-84-8564  S99994136  X23175682X    Mr.      Abram53        Cole117   
3     999-54-8345  S99917811  X33252218X   Mrs.   Eufemia350  Jakubowski832   
4     999-71-9025  S99938018  X79840478X    Mr.     Javier97       Terán294   
...           ...        ...         ...    ...          ...            ...   
5710  999-22-2904  S99962544  X63396209X    Mr.   Octavio643       Muñoz828   
5711  999-71-2826  S99958264  X62286915X    Mr.       Abe604       Stark857   
5712  999-97-2074  S99956339  X38583829X    Mr.   Forrest301      Waters156   
5713  999-20-9898  S99913259  X87294055X    Mr.     Kelly223    Ondricka197   
5714  999-12-6776  S99914184  X77226361X   Mrs.  Graciela518       Zarate81   

     SUFFIX  ...           CITY       STATE                COUNTY    FIPS  \
0       NaN  ...         Walnut  California    Los Angeles County  6037.0   
1       NaN  ...        Soledad  California       Monterey County  6053.0   
2       NaN  ...        Hanford  California          Kings County  6031.0   
3       NaN  ...  San Francisco  California  San Francisco County  6075.0   
4       NaN  ...        Arcadia  California    Los Angeles County  6037.0   
...     ...  ...            ...         ...                   ...     ...   
5710    NaN  ...      San Diego  California      San Diego County  6073.0   
5711    NaN  ...     Costa Mesa  California         Orange County  6059.0   
5712    NaN  ...        Norwalk  California    Los Angeles County  6037.0   
5713    NaN  ...     Costa Mesa  California         Orange County  6059.0   
5714    NaN  ...         Carson  California    Los Angeles County  6037.0   

        ZIP        LAT         LON HEALTHCARE_EXPENSES HEALTHCARE_COVERAGE  \
0     91724  34.110204 -117.803515            25831.61             3234.86   
1     93960  36.342698 -121.434998            36690.25                0.00   
2     93230  36.260623 -119.615211             6289.69            54030.33   
3     94111  37.780116 -122.450481            16832.02          2339850.05   
4     91732  34.138577 -118.065934            66081.73            20372.03   
...     ...        ...         ...                 ...                 ...   
5710  92122  32.901176 -117.038117            60885.15            14885.42   
5711  92626  33.631114 -117.798555           152505.53          2443760.53   
5712  90703  33.836854 -118.078111            13482.06          1047683.40   
5713  92646  33.724060 -117.903008           354873.71          1065469.66   
5714  90248  33.868254 -118.249119           737028.50           851526.10   

      INCOME  
0     125285  
1      47629  
2      10371  
3      19571  
4     135002  
...      ...  
5710  894206  
5711  106186  
5712   87279  
5713  106186  
5714   88949  

[5715 rows x 27 columns]

In [ ]:
conditions = pd.read_csv("data/conditions.csv")
print(conditions.columns)
#select patients with anxiety
conditions = conditions[conditions["DESCRIPTION"].str.contains("anxiety")]
conditions

In [ ]:
anxiety_patients = pd.merge(patients, conditions, left_on='Id', right_on='PATIENT')
#select specific columns
anxiety_patients = anxiety_patients[['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'DESCRIPTION']]
anxiety_patients

NOTE: Multiple observations for same patient can exist 

In [8]:
observations = pd.read_csv("data/observations.csv")
print(observations.columns)
#select observations recorded for anxiety
observations = observations[observations["DESCRIPTION"].str.contains("anxiety")]


Index(['DATE', 'PATIENT', 'ENCOUNTER', 'CATEGORY', 'CODE', 'DESCRIPTION',
       'VALUE', 'UNITS', 'TYPE'],
      dtype='object')


In [ ]:
anxiety_observations = pd.merge(anxiety_patients, observations, left_on='Id', right_on='PATIENT')
#select columns of interest
anxiety_observations = anxiety_observations[['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME', 'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'DESCRIPTION_x', 'DESCRIPTION_y', 'VALUE']]
anxiety_observations

In [ ]:
anxiety_observations = anxiety_observations.rename(columns={'DESCRIPTION_x':'CONDITION', 'DESCRIPTION_y':'OBSERVED_CONDITION'})
anxiety_observations

Data cleaning on Synthetic medical data

Drop rows with ZIP = 0

In [43]:
anxiety_df = anxiety_observations[anxiety_observations.ZIP != 0]
anxiety_df.shape

In [46]:
# Number of cities in California
anxiety_df.groupby('ZIP').size()

ZIP
90001    208
90002      3
90005      4
90006     31
90007      9
        ... 
95949      3
95963     21
95971     76
96003     13
96117     40
Length: 460, dtype: int64

In [90]:
# Number of California CITY (inside COUNTY) 
print(patients[patients['ZIP'] != 0].groupby('CITY').size())
# print(patients[patients['ZIP'] != 0].groupby('COUNTY').size())

CITY
Acton            2
Adelanto         7
Agoura Hills     2
Alameda         10
Alamo            1
                ..
Yountville       3
Yreka            2
Yuba City       16
Yucaipa          9
Yucca Valley     8
Length: 574, dtype: int64


In [77]:
anxiety_observations.columns

Index(['Id', 'ZIP', 'RACE', 'GENDER', 'ETHNICITY', 'INCOME',
       'HEALTHCARE_EXPENSES', 'HEALTHCARE_COVERAGE', 'CONDITION',
       'OBSERVED_CONDITION', 'VALUE'],
      dtype='object')

In the **anxiety_df dataset** of California state in US, there are 460 city/zip codes. In the **vic_zipcode dataset** of Victoria state in Australia, there are 749 suburbs/zip code.

In [122]:
# Paste your own full file path
aus_zipcode = pd.read_csv("/Users/cindy_homefolder/Desktop/y1s2/ENGR90033/choropleth map/synthea 2023/australian data/zipcodes_au.csv", index_col=None)
aus_zipcode.shape

(18513, 6)

In [123]:
aus_zipcode

USPS   ST                NAME  ZCTA5  \
0      Australian Capital Territory  ACT  Unincorporated ACT    200   
1      Australian Capital Territory  ACT  Unincorporated ACT    200   
2      Australian Capital Territory  ACT  Unincorporated ACT   2540   
3      Australian Capital Territory  ACT  Unincorporated ACT   2540   
4      Australian Capital Territory  ACT  Unincorporated ACT   2540   
...                             ...  ...                 ...    ...   
18508             Western Australia   WA            Gosnells   6989   
18509             Western Australia   WA            Gosnells   6990   
18510             Western Australia   WA            Armadale   6991   
18511             Western Australia   WA       Victoria Park   6992   
18512             Western Australia   WA            Armadale   6997   

             LAT         LON  
0     -35.277700  149.119000  
1     -35.277700  149.118900  
2     -34.863483  150.616440  
3     -34.863483  150.616440  
4     -35.129878  150.707521  
...          ...         ...  
18508 -32.039597  116.009575  
18509 -32.060407  116.008461  
18510 -32.097880  116.095955  
18511 -31.964503  115.920395  
18512 -32.097880  116.095955  

[18513 rows x 6 columns]

Generate Victoria dataset and replace ZIP column in anxiety_df to Victoria postcode due to the unsolved custom module issue of Synthea. 

In [124]:
vic_zipcode = aus_zipcode[aus_zipcode['USPS'] == 'Victoria']
vic_zipcode.shape


(3540, 6)

In [125]:
vic_zipcode

USPS   ST               NAME  ZCTA5        LAT         LON
13034  Victoria  VIC          Melbourne   3000 -37.817403  144.956776
13035  Victoria  VIC          Melbourne   3001 -37.817403  144.956776
13036  Victoria  VIC              Yarra   3002 -37.818517  144.982207
13037  Victoria  VIC          Melbourne   3003 -37.810871  144.949592
13038  Victoria  VIC       Port Phillip   3004 -37.844246  144.970161
...         ...  ...                ...    ...        ...         ...
16569  Victoria  VIC    South Gippsland   8438   0.000000    0.000000
16570  Victoria  VIC    South Gippsland   8511   0.000000    0.000000
16571  Victoria  VIC  Greater Dandenong   8785 -38.016114  145.208504
16572  Victoria  VIC  Greater Dandenong   8785 -38.016114  145.208504
16573  Victoria  VIC          Melbourne   9999 -37.817403  144.956776

[3540 rows x 6 columns]

The dataset got no suburb level, only the state level in zip code column, ZCTA5. Therefore, multiple zip codes belong to one suburb. 

In [126]:
vic_zipcode.groupby('NAME').size()

NAME
Alpine           48
Ararat           40
Ballarat        105
Banyule          25
Bass Coast       71
               ... 
Wodonga          15
Wyndham          27
Yarra            26
Yarra Ranges     29
Yarriambiack     25
Length: 80, dtype: int64

In [127]:
vic_zipcode.groupby('ZCTA5').size()

ZCTA5
3000    1
3001    1
3002    1
3003    1
3004    3
       ..
8205    1
8438    1
8511    1
8785    2
9999    1
Length: 749, dtype: int64

In [128]:
# vic_zipcode[vic_zipcode.ZCTA5 == 9999]
# vic_zipcode[vic_zipcode.NAME == 'Yarra']

Consider the data distribution in two datasets and their different dimension, I will take SEIFA data into consider too. 

In [129]:
demog = pd.read_csv("/Users/cindy_homefolder/Desktop/y1s2/ENGR90033/choropleth map/synthea 2023/australian data/demographics_au.csv")
demog.head()

ID  COUNTY               NAME           STNAME  POPESTIMATE2015  \
0  10050   10050             Albury  New South Wales            56036   
1  10180   10180  Armidale Regional  New South Wales            29332   
2  10250   10250            Ballina  New South Wales            46172   
3  10300   10300          Balranald  New South Wales             2207   
4  10470   10470  Bathurst Regional  New South Wales            43653   

             CTYNAME  TOT_POP  TOT_MALE  TOT_FEMALE     WHITE  ...  35..50  \
0             Albury    56036  0.487401    0.512599  0.522188  ...  0.0980   
1  Armidale Regional    29332  0.485068    0.514932  0.522188  ...  0.1096   
2            Ballina    46172  0.486529    0.513471  0.522188  ...  0.1108   
3          Balranald     2207  0.496602    0.503398  0.522188  ...  0.1156   
4  Bathurst Regional    43653  0.507594    0.492406  0.522188  ...  0.1192   

   50..75  75..100  100..150  150..200  200..999  LESS_THAN_HS  HS_DEGREE  \
0  0.0976   0.0976    0.1008    0.1040    0.1040      0.405605   0.171377   
1  0.1120   0.1120    0.0956    0.0792    0.0792      0.405605   0.171377   
2  0.0924   0.0924    0.0826    0.0728    0.0728      0.405605   0.171377   
3  0.0908   0.0908    0.0828    0.0748    0.0748      0.405605   0.171377   
4  0.1164   0.1164    0.0870    0.0576    0.0576      0.405605   0.171377   

   SOME_COLLEGE  BS_DEGREE  
0      0.106205   0.316813  
1      0.106205   0.316813  
2      0.106205   0.316813  
3      0.106205   0.316813  
4      0.106205   0.316813  

[5 rows x 47 columns]

In [130]:
demog.groupby('CTYNAME').size()

CTYNAME
Adelaide           1
Adelaide Hills     1
Adelaide Plains    1
Albany             1
Albury             1
                  ..
Yarriambiack       1
Yass Valley        1
Yilgarn            1
York               1
Yorke Peninsula    1
Length: 546, dtype: int64

Take city as a common column, filter the common city that occurs in both datasets, victoria zipcode and anxiety df, then spread the postcode by ID column in anxiety datasets.

In [131]:
# vic_zipcode['NAME'].isin(demog['CTYNAME']).value_counts() #True : 3413 has repeated values
demog['CTYNAME'].isin(vic_zipcode['NAME']).value_counts()

False    469
True      77
Name: CTYNAME, dtype: int64

In [ ]:
# demog got no postcode column, so I have to use vic_zipcode
# demog['CTYNAME'].loc[demog['CTYNAME'].isin(vic_zipcode['NAME'])]
demog[demog['CTYNAME'].isin(vic_zipcode['NAME'])]

In [138]:
unified_vic_zipcode = vic_zipcode[vic_zipcode['NAME'].isin(demog['CTYNAME'])]

In [144]:
unified_vic_zipcode.shape

(3413, 6)

In [143]:
len(unified_vic_zipcode.ZCTA5.unique())

722

In [150]:
len(anxiety_df.groupby('Id').size())

664

**664 patients's postcode need to be replaced using 722 available postcodes in Victoria.**

In [ ]:
unified_vic_zipcode.groupby('ZCTA5').size()

In [159]:
# Create a mapping between 'Id' and 'ZIP'
id_to_zip = dict(zip(anxiety_df['Id'].unique(), unified_vic_zipcode['ZCTA5']))

# Use the map function to replace 'ZIP' values based on 'Id'
anxiety_df['ZIP'] = anxiety_df['Id'].map(id_to_zip)

/var/folders/vt/3bwlqrgx039fhm0gqgx_mlyc0000gn/T/ipykernel_14870/1593299730.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  anxiety_df['ZIP'] = anxiety_df['Id'].map(id_to_zip)


### After replacement, let's have a look at the new dataset.

In [171]:
anxiety_df.shape

(14334, 11)

In [169]:
anxiety_df.head()

Id   ZIP   RACE GENDER ETHNICITY  INCOME  \
0  16c73553-9400-1f2c-bcf9-7b5ef75150d2  3000  white      F  hispanic   45504   
1  16c73553-9400-1f2c-bcf9-7b5ef75150d2  3000  white      F  hispanic   45504   
2  16c73553-9400-1f2c-bcf9-7b5ef75150d2  3000  white      F  hispanic   45504   
3  16c73553-9400-1f2c-bcf9-7b5ef75150d2  3000  white      F  hispanic   45504   
4  16c73553-9400-1f2c-bcf9-7b5ef75150d2  3000  white      F  hispanic   45504   

   HEALTHCARE_EXPENSES  HEALTHCARE_COVERAGE                         CONDITION  \
0            286834.48            130435.94  Severe anxiety (panic) (finding)   
1            286834.48            130435.94  Severe anxiety (panic) (finding)   
2            286834.48            130435.94  Severe anxiety (panic) (finding)   
3            286834.48            130435.94  Severe anxiety (panic) (finding)   
4            286834.48            130435.94  Severe anxiety (panic) (finding)   

                                  OBSERVED_CONDITION VALUE  
0  Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
1  Generalized anxiety disorder 7 item (GAD-7) to...   3.0  
2  Generalized anxiety disorder 7 item (GAD-7) to...  21.0  
3  Generalized anxiety disorder 7 item (GAD-7) to...   2.0  
4  Generalized anxiety disorder 7 item (GAD-7) to...   3.0

In [161]:
len(anxiety_df.ZIP.unique())

212

In [166]:
anxiety_df.groupby('ZIP').size()

ZIP
3000      7
3001      3
3002     15
3003     11
3004     57
       ... 
3249    475
3250     87
3251     47
3254      3
3260    372
Length: 212, dtype: int64

In [168]:
# Check number of patients in one zip code
anxiety_df[anxiety_df.ZIP == 3260].Id.nunique()

15

In [164]:
# Having the same number of patients
anxiety_df.groupby(['Id', 'ZIP']).size()

Id                                    ZIP 
012aea97-b365-89a3-9a5e-f01882baf2d9  3065    50
0135a402-de4f-03f7-9b57-b21235269361  3161     5
0141371c-5848-c0d1-153b-29ec44de0561  3072     2
02852ee9-4629-e1f6-116f-fa52d77f2845  3216    15
028b01e2-02dc-e79e-54e3-856f9d3a8ed7  3103    18
                                              ..
fceacc48-c665-c4e2-ec22-e3d589336a4e  3040     9
fe6df80f-acff-34b4-3f60-f9f4f09e8a1c  3204     1
fe6ec497-bb8d-f614-5bcd-25324a5edd53  3121    56
feeb4853-6336-a868-130f-b733e4a17aca  3061     2
fffe51a2-5f76-d521-6079-bce7b606b42e  3207     4
Length: 664, dtype: int64

In [170]:
anxiety_df.to_csv("vic_anxiety.csv", encoding='utf-8', index=False)